### ISSUE TITLE
transformer complains about CSS

### ISSUE THREAD

**Issue by zoechi** [dart-lang/angular2#330]
since updating to newest dependencies today 

```
angular2 3.0.0-beta
```
and some other updates

```
 (master) $ pub upgrade --packages-dir
Resolving dependencies... (39.9s)
  analyzer 0.29.10 (0.30.0-alpha.1 available)
> angular2 3.0.0-beta (was 3.0.0-alpha+1)
> angular_test 1.0.0-beta+2 (was 1.0.0-beta+1)
> code_builder 1.0.1 (was 1.0.0)
> dart_style 1.0.1 (was 0.2.16) (1.0.2-dev available)
> observable 0.20.4+1 (was 0.20.4)
```

```
[web]  GET Served 23 cached assets.
[Warning from DirectiveMetadataLinker on fhir_webui|lib/src/webui_generated/form/edit_mode_service.ng_summary.json]:
Errors parsing CSS:
error on line 7000, column 11: expected :, but found (
@supports (-webkit-appearance: none) {
          ^
error on line 7000, column 36: expected {, but found )
@supports (-webkit-appearance: none) {
                                   ^
error on line 7000, column 38: expected }, but found {
@supports (-webkit-appearance: none) {
                                     ^
error on line 7004, column 46: premature end of file unknown CSS
    mask: url("/images/buffer.svg?embed"); } }
                                             ^
```

**Comment by jimmyff** [comment]
I also have this issue, can confirm that rolling back to 3.0.0-alpha+1 resolved the issue for me.

Here is the build error:
```
Build completed with 1 errors.
Build error:
Transform StylesheetCompiler on dn_angular|lib/components/profile_summaries/profile_summaries_component.css threw error: The getter 'text' was called on null.
Receiver: null
Tried calling: text
dart:core-patch/object_patch.dart 44                                  Object._noSuchMethod
dart:core-patch/object_patch.dart 47                                  Object.noSuchMethod
package:csslib/parser.dart 2370                                       _Parser.processDimension
package:csslib/parser.dart 2326                                       _Parser.processTerm
package:csslib/parser.dart 2078                                       _Parser.processExpr
package:csslib/parser.dart 1687                                       _Parser.processDeclaration
package:csslib/parser.dart 1067                                       _Parser.processDeclarations
package:csslib/parser.dart 996                                        _Parser.processRuleSet
package:csslib/parser.dart 529                                        _Parser.processDirective
package:csslib/parser.dart 208                                        _Parser.parse
package:csslib/parser.dart 97                                         parse
package:angular2/src/compiler/shadow_css.dart 100                     shimShadowCss
package:angular2/src/compiler/style_compiler.dart 97                  StyleCompiler._shimIfNeeded
package:angular2/src/compiler/style_compiler.dart 73                  StyleCompiler._compileStyles
package:angular2/src/compiler/style_compiler.dart 64                  StyleCompiler.compileStylesheet
package:angular2/src/compiler/offline_compiler.dart 119               OfflineCompiler.compileStylesheet
package:angular2/src/transform/stylesheet_compiler/processor.dart 32  processStylesheet.<async>.<fn>.<async>
dart:async/future.dart 184                                            Future.Future.microtask.<fn>
dart:async/zone.dart 1150                                             _rootRun
dart:async/zone.dart 1026                                             _CustomZone.run
dart:async/zone.dart 953                                              _CustomZone.bindCallback.<fn>
dart:async/schedule_microtask.dart 41                                 _microtaskLoop
dart:async/schedule_microtask.dart 50                                 _startMicrotaskLoop
dart:isolate-patch/isolate_patch.dart 96                              _runPendingImmediateCallback
dart:isolate-patch/isolate_patch.dart 149                             _RawReceivePortImpl._handleMessage

dart:core                                                             Object.noSuchMethod
package:csslib/parser.dart 2370                                       _Parser.processDimension
package:csslib/parser.dart 2326                                       _Parser.processTerm
package:csslib/parser.dart 2078                                       _Parser.processExpr
package:csslib/parser.dart 1687                                       _Parser.processDeclaration
package:csslib/parser.dart 1067                                       _Parser.processDeclarations
package:csslib/parser.dart 996                                        _Parser.processRuleSet
package:csslib/parser.dart 529                                        _Parser.processDirective
package:csslib/parser.dart 208                                        _Parser.parse
package:csslib/parser.dart 97                                         parse
package:angular2/src/compiler/shadow_css.dart 100                     shimShadowCss
package:angular2/src/compiler/style_compiler.dart 97                  StyleCompiler._shimIfNeeded
package:angular2/src/compiler/style_compiler.dart 73                  StyleCompiler._compileStyles
package:angular2/src/compiler/style_compiler.dart 64                  StyleCompiler.compileStylesheet
package:angular2/src/compiler/offline_compiler.dart 119               OfflineCompiler.compileStylesheet
package:angular2/src/transform/stylesheet_compiler/processor.dart 32  processStylesheet.<async>.<fn>.<async>
```

I tracked it down to a line in my sass code:
```
width: $card-width-percent;
````

This variable was being set here:
```
$card-width-percent: percentage(1 / $i);
```

**Comment by natebosch** [comment]
@leonsenft is this invalid syntax that is now rejected by the stricter parser?

**Comment by matanlurey** [comment]
I thought we had a flag to opt-out of the new parser too? @leonsenft

**Comment by leonsenft** [comment]
Sorry, this is an issue with the [CSS parser](https://github.com/dart-lang/csslib) we're using. I'll add support for @supports and resolve this issue when I have.

@matanlurey Unfortunately the flag doesn't opt out of using the new parser, it just enables the old behavior for shadow host selectors.

**Comment by leonsenft** [comment]
@jimmyff that looks like a different issue. Could you please provide me with the full SCSS source for me to investigate?

**Comment by ligustah** [comment]
I'm using [MDL](https://getmdl.io) (still using version 1.2.1, though) and get errors when running `build` or `serve`, which for me is particularly annoying, because it'll try to print the offending line to the console. For the minified mdl css files that means basically printing the complete file to stderr for every error that it finds. It also appears to load all css files that is finds, because removing the reference to the minified file still showed the same behavior. Only renaming the file stopped it.

This is not affecting me in any serious way since it's just a toy project anyways. I just wanted to share my experience.

**Comment by leonsenft** [comment]
@ligustah Could you please tell me which file it is so I can investigate the error? Thanks.

**Comment by ligustah** [comment]
@leonsenft Both material.css and material.min.css from this release of mdl: https://github.com/google/material-design-lite/releases/tag/v1.2.1

Thanks for looking into it!
(though as I said, it's not really a blocker for me right now)

**Comment by leonsenft** [comment]
@zoechi @ligustah I just published a new version of [csslib](https://pub.dartlang.org/packages/csslib) which adds supports for `@supports`. Run `pub upgrade`, verify you're on csslib 0.13.5, and this issue should be fixed.

**Comment by ligustah** [comment]
I just tried it with `angular2 3.0.0-beta` and `csslib 0.13.5` and haven't seen an error. Thanks a lot for the quick fix!

**Comment by geekrumper** [comment]
Hi, 
are there any new information about the "opt-out" feature? Is it available? How can I use it? I'm using a third party css library [semantic-ui](https://github.com/Semantic-Org/Semantic-UI/blob/master/dist/semantic.css) and the StyleSheetCompiler is still complaining. It's also very time consuming to get both parties integrated (semantic-ui | csslib).

**Comment by leonsenft** [comment]
Hi @geekrumper,

The opt-out feature is unrelated to the issues you're seeing with Semantic UI. You're right the example you've shown is absolutely correct CSS. Unfortunately the CSS parser we adopted is proving to be far less spec-compliant than initially believed. I will resolve these issues with the CSS parser, but in the meantime there's no way to opt-out of using it. If this issue is blocking you, you'll have to revert to a pre-3.0.0 version of Angular temporarily, or modify the CSS. I know neither of these solutions are very ideal, and I apologize for how disruptive the CSS changes have been.

Along with making the parser spec compliant, I'm also going to improve error recovery so that if a single selector is problematic, it doesn't prevent the remainder of the CSS from being shimmed.

In case you're curious, the opt-out feature refers to the behavior of the shadow host selectors `:host`, `:host()`, and `:host-context()`. There was a long standing bug in Angular where these selectors prevented following selectors from being properly scoped to their component, thus violating style encapsulation. While the correct syntax to pierce shadow boundaries now is to use a shadow piercing combinator `>>>` or `/deep/`, these combinators weren't always supported by Angular, so the aforementioned bug was widely used in their place.

We wanted to fix this bug, but acknowledged how many existing styles it could break, so we've added a transformer option `use_legacy_style_encapsulation` to enable the old, bugged, behavior. In your *pubspec.yaml*, you can enable this option like so:

```yaml
...
transformers:
  - angular2:
      use_legacy_style_encapsulation: true
...
```

Again I'm sorry you're experiencing these CSS issues. I'll update you as soon as they're resolved.

**Comment by geekrumper** [comment]
Hello @leonsenft,
very good information!!! Thank you. Using `3.0.0-alpha+1` now.

**Comment by kevmoo** [comment]
> very good information!!! Thank you. Using 3.0.0-alpha+1 now.

You should upgrade to `3.0.0` gold, @geekrumper 😄

**Comment by geekrumper** [comment]
> You should upgrade to 3.0.0 gold, @geekrumper 😄

@kevmoo, every version between `3.0.0-beta` and `3.0.0` included is having this issue.

**Comment by Charles-LMQ** [comment]
since I upgraded to 3.1.0,  the css parsing get worse than before...

Full screen filled:

`error on line 1, column 4597: expected }, but found =`


while they are just quite normal js-plugins, such like:

css/select2.min.css

**Comment by zoechi** [comment]
@Charles-LMQ CSS doesn't cause issues for me.
I guess more information is required to diagnose your problem.

**Comment by Charles-LMQ** [comment]
@zoechi yeah, I want, but the screen flash so fast, that I only catch some keyword here:

```
package:$pub/serialize/transformer.dart 17:5                                _serializeTransformer.<fn>
[DirectiveProcessor on cms|lib/ui/home/HomeComponent.dart]:
Errors parsing CSS:
error on line 5, column 108407: expected { after media before ruleset

```

and then the console is filled by:

[https://contentplatform-dev-main-44062.s3-ap-southeast-1.amazonaws.com/content_thumbnail/dartcsspro_1499417854265.PNG](url)

**Comment by leonsenft** [comment]
@Charles-LMQ If you could please post a link to the source or actual text rather than a screenshot so I can copy and debug that would be appreciated. Thanks!

**Comment by Charles-LMQ** [comment]
@leonsenft thanks a lot, dude, but we already removeed those libs in our project, which we decide not to use them, but we recorded their names: select2, jquery-ui which contains a lot of  "@import", hope it is useful.

**Comment by leonsenft** [comment]
@Charles-LMQ Thanks for the heads up. Was the issue by any chance https://github.com/dart-lang/angular/issues/39? If so this should be fixed in 4.0.0 when it's published.

**Comment by Charles-LMQ** [comment]
@leonsenft hi, yeah, it is. OK, sure, no worries anymore.
